collector/mountstats_linux.go

```diff
@@ -507,37 +508,51 @@ func (c *mountStatsCollector) Update(ch chan<- prometheus.Metric) error {
 		return fmt.Errorf("failed to parse mountstats: %v", err)
 	}
 
+	mountsInfo, err := c.proc.MountInfo()
+	if err != nil {
+		return fmt.Errorf("failed to parse mountinfo: %v", err)
+	}
+
 	// store all seen nfsDeviceIdentifiers for deduplication
 	deviceList := make(map[nfsDeviceIdentifier]bool)
 
-	for _, m := range mounts {
+	for idx, m := range mounts {
 		// For the time being, only NFS statistics are available via this mechanism
 		stats, ok := m.Stats.(*procfs.MountStatsNFS)
+
 		if !ok {
 			continue
 		}
 
-		deviceIdentifier := nfsDeviceIdentifier{m.Device, stats.Transport.Protocol}
+		var mountAddress string
+		if idx < len(mountsInfo) {
+			// The mount entry order in the /proc/self/mountstats and /proc/self/mountinfo is the same.
+			miStats := mountsInfo[idx]
+			mountAddress = miStats.SuperOptions["addr"]
+		}
+
+		deviceIdentifier := nfsDeviceIdentifier{m.Device, stats.Transport.Protocol, mountAddress}
 		i := deviceList[deviceIdentifier]
 		if i {
 			log.Debugf("Skipping duplicate device entry %q", deviceIdentifier)
 			continue
 		}
 
 		deviceList[deviceIdentifier] = true
-		c.updateNFSStats(ch, m.Device, stats.Transport.Protocol, stats)
+		c.updateNFSStats(ch, m.Device, stats.Transport.Protocol, mountAddress, stats)
 	}
 
 	return nil
 }
 
-func (c *mountStatsCollector) updateNFSStats(ch chan<- prometheus.Metric, export string, protocol string, s *procfs.MountStatsNFS) {
+func (c *mountStatsCollector) updateNFSStats(ch chan<- prometheus.Metric, export string, protocol string, mountAddress string, s *procfs.MountStatsNFS) {
 	ch <- prometheus.MustNewConstMetric(
 		c.NFSAgeSecondsTotal,
 		prometheus.CounterValue,
 		s.Age.Seconds(),
 		export,
 		protocol,
+		mountAddress,
```

Comment:
To save some lines of repetition here maybe a vararg slice would be better instead of the individual strings.
```golang
	labelValues := []string{export, protocol, mountAddress}
	ch <- prometheus.MustNewConstMetric(
		c.NFSAgeSecondsTotal,
		prometheus.CounterValue,
		s.Age.Seconds(),
		labelValues...,
```

Reply:
We could probably just pass in the vararg slice to the `updateNFSStats` function itself!

Reply:
Or better yet, pass in the `nfsDeviceIdentifier` struct instance.

Reply:
```
for _, op := range s.Operations {
ch <- prometheus.MustNewConstMetric(
	c.NFSOperationsRequestsTotal,
	prometheus.CounterValue,
	float64(op.Requests),
	labelValues...,
        op.Operation,
)
...
```
This does not seem to work using slice expansion.